plugins/inputs/zipkin/convert.go

```diff
@@ -0,0 +1,245 @@
+			BinaryAnnotations: NewBinaryAnnotations(span.GetBinaryAnnotations(), endpoint),
+		}
+	}
+	return trace
+}
+
+// NewAnnotations converts a slice of *zipkincore.Annotation into a slice
+// of new Annotations
+func NewAnnotations(annotations []*zipkincore.Annotation, endpoint *zipkincore.Endpoint) []Annotation {
+	formatted := make([]Annotation, len(annotations))
+	for i, annotation := range annotations {
+		formatted[i] = Annotation{
+			Host:        host(endpoint),
+			ServiceName: serviceName(endpoint),
+			Timestamp:   microToTime(annotation.GetTimestamp()),
+			Value:       annotation.GetValue(),
+		}
+	}
+
+	return formatted
+}
+
+// NewBinaryAnnotations is very similar to NewAnnotations, but it
+// converts zipkincore.BinaryAnnotations instead of the normal zipkincore.Annotation
+func NewBinaryAnnotations(annotations []*zipkincore.BinaryAnnotation, endpoint *zipkincore.Endpoint) []BinaryAnnotation {
+	formatted := make([]BinaryAnnotation, len(annotations))
+	for i, annotation := range annotations {
+		formatted[i] = BinaryAnnotation{
+			Host:        host(endpoint),
+			ServiceName: serviceName(endpoint),
+			Key:         annotation.GetKey(),
+			Value:       string(annotation.GetValue()),
+			Type:        annotation.GetAnnotationType().String(),
+		}
+	}
+	return formatted
+}
+
+func microToTime(micro int64) time.Time {
+	return time.Unix(0, micro*int64(time.Microsecond)).UTC()
+}
+
+func formatID(id int64) string {
+	return strconv.FormatInt(id, 10)
+}
+
+func formatTraceID(high, low int64) string {
+	if high == 0 {
+		return fmt.Sprintf("%x", low)
+	}
+	return fmt.Sprintf("%x%016x", high, low)
+}
+
+func minMax(span *zipkincore.Span) (time.Time, time.Time) {
+	min := now().UTC()
+	max := time.Time{}.UTC()
+	for _, annotation := range span.Annotations {
+		ts := microToTime(annotation.GetTimestamp())
+		if !ts.IsZero() && ts.Before(min) {
+			min = ts
+		}
+		if !ts.IsZero() && ts.After(max) {
+			max = ts
+		}
+	}
+	if max.IsZero() {
+		max = min
+	}
+	return min, max
+}
+
+func guessTimestamp(span *zipkincore.Span) time.Time {
+	if span.GetTimestamp() != 0 {
+		return microToTime(span.GetTimestamp())
+	}
+	min, _ := minMax(span)
+	return min
+}
+
+func convertDuration(span *zipkincore.Span) time.Duration {
+	duration := time.Duration(span.GetDuration()) * time.Microsecond
+	if duration != 0 {
+		return duration
+	}
+	min, max := minMax(span)
+	return max.Sub(min)
+}
+
+func parentID(span *zipkincore.Span) string {
+	// A parent ID of 0 means that this is a parent span. In this case,
+	// we set the parent ID of the span to be its own id, so it points to
+	// itself.
+	id := span.GetParentID()
+	if id != 0 {
+		return formatID(id)
+	}
+	return formatID(span.ID)
+}
+
+func ipv4(addr int32) string {
+	buf := make([]byte, 4)
+	binary.BigEndian.PutUint32(buf, uint32(addr))
+	return net.IP(buf).String()
+}
+
+func host(h *zipkincore.Endpoint) string {
+	if h == nil {
+		return ""
```

Comment:
Return a default non-empty string value or guard against adding this in the recorder